### PR TITLE
MDEXP-621 - Export all endpoint

### PR DESCRIPTION
**Pull request by obozhko-folio**
[MDEXP-621](https://issues.folio.org/browse/MDEXP-621) - Export all endpoint

## Purpose
To provide API for exporting the entire collection.

## Approach
Use slicer component.

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] Check logging
  - [x] There are no major code smells or security issues

- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/dataexp/service/export/strategies/HoldingsExportStrategy.java [hunk 1]
```diff
@@ -83,10 +113,29 @@ public Optional<String> getIdentifierMessage(UUID id) {
...
   }
 
-  protected List<JSONObject> getHoldingsWithInstanceAndItems(Set<UUID> holdingsIds, GeneratedMarcResult result, MappingProfile mappingProfile) {
+  @Override
+  public Map<UUID,MarcFields> getAdditionalMarcFieldsByExternalId(List<MarcRecordEntity> marcRecords, MappingProfile mappingProfile) {
+    return new HashMap<>();
+  }
+
+  protected List<JSONObject> getHoldingsWithInstanceAndItems(Set<UUID> holdingsIds, GeneratedMarcResult result,
+      MappingProfile mappingProfile, ExportRequest exportRequest, boolean lastSlice, boolean lastExport) {
```
Comment (alekGbuz) [inline src/main/java/org/folio/dataexp/service/export/strategies/HoldingsExportStrategy.java:122]: Hi @obozhko-folio , could please a bit rework usage of boolean lastSlice, boolean lastExport arguments as not clear the usage of this flags.

File: src/main/java/org/folio/dataexp/service/export/ExportContext.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class ExportContext {
+
+  private final AtomicBoolean lastSlice;
+  private final AtomicBoolean lastExport;
+
+  public void setLastSlice(boolean lastSlice) {
+    this.lastSlice.set(lastSlice);
```
Comment (alekGbuz) [inline src/main/java/org/folio/dataexp/service/export/ExportContext.java:16]: hi @obozhko-folio, as ExportContext is singleton here - setters could produce unexpected behavior for execution in several threads.

File: src/main/java/org/folio/dataexp/service/export/strategies/InstancesExportAllStrategy.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    updateSliceState(slice);
+    log.info("Slice size for instances export all: {}", slice.getSize());
+    var exportIds = slice.getContent().stream().map(InstanceEntity::getId).collect(Collectors.toSet());
+    log.info("Size of exportIds for instances export all: {}", exportIds.size());
+    createAndSaveMarc(exportIds, exportStatistic, mappingProfile, exportFilesEntity.getJobExecutionId(),
+      exportRequest);
+    while (slice.hasNext()) {
+      slice = chooseSlice(exportFilesEntity, exportRequest, slice.nextPageable());
+      updateSliceState(slice);
+      exportIds = slice.getContent().stream().map(InstanceEntity::getId).collect(Collectors.toSet());
```
Comment (alekGbuz) [inline src/main/java/org/folio/dataexp/service/export/strategies/InstancesExportAllStrategy.java:74]: @obozhko-folio, seems here only ids of instances are needed so it is possible to reduce database load if retrieve only ids . Could you please double check this.

**Review by alekGbuz — APPROVED** [review]

File: src/main/java/org/folio/dataexp/repository/MarcInstanceRecordRepository.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import jakarta.persistence.PersistenceContext;
+import jakarta.persistence.Query;
+import org.folio.dataexp.domain.entity.MarcRecordEntity;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+
+import static java.lang.String.format;
```
Comment (siarhei-charniak) [inline src/main/java/org/folio/dataexp/repository/MarcInstanceRecordRepository.java:13]: Please move static import to the top

File: src/main/java/org/folio/dataexp/service/DataExportAllService.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import org.folio.dataexp.domain.dto.ExportAllRequest;
+import org.folio.dataexp.domain.dto.ExportRequest;
+import org.folio.dataexp.domain.dto.FileDefinition;
+import org.folio.dataexp.repository.JobProfileEntityRepository;
+import org.springframework.stereotype.Service;
+
+import java.util.UUID;
+
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
```
Comment (siarhei-charniak) [inline src/main/java/org/folio/dataexp/service/DataExportAllService.java:14]: Please move static imports to the top

**Review by siarhei-charniak — APPROVED** [review]

File: src/main/java/org/folio/dataexp/repository/MarcInstanceRecordRepository.java (hunk 4 above)
Comment (obozhko-folio) [reply src/main/java/org/folio/dataexp/repository/MarcInstanceRecordRepository.java]: Done.

File: src/main/java/org/folio/dataexp/service/DataExportAllService.java (hunk 5 above)
Comment (obozhko-folio) [reply src/main/java/org/folio/dataexp/service/DataExportAllService.java]: Done.

**Pull request merged**